src/fpm_targets.f90

```diff
@@ -473,14 +473,22 @@ end subroutine add_dependency
 !> a source file in the package of the correct scope, then a __fatal error__
 !> is returned by the procedure and model construction fails.
 !>
-subroutine resolve_module_dependencies(targets,external_modules,error)
+subroutine resolve_module_dependencies(targets,external_modules,source_dirs,error)
     type(build_target_ptr), intent(inout), target :: targets(:)
```

Comment:
The logic of adding `source_dirs` argument breaks the fpm source file scoping specification. For details, see:
1. [Source file scoping](https://fortran-lang.github.io/fpm/proc/resolve_module_dependencies.html)
2. https://github.com/fortran-lang/fpm-docs/issues/95

There may be some discussion or improvement needed to the source file scope of `app`, `example`, and `test`.

Reply:
@zoziha Thank you for the thorough review - this was not entirely clear to me when I read the code, thank you for documenting it. I will work on some modifications, because like you're pointing out, right now a unique directory search list is built for all targets